language/functional_tests/src/checker/directives.rs

```diff
@@ -0,0 +1,204 @@
+                        Some('t') => buffer.push('\t'),
+                        Some('r') => buffer.push('\r'),
+                        Some('"') => buffer.push('"'),
+                        Some(c) => bail!("unrecognized escape character \\{}", c),
+                        None => bail!("unclosed escape character"),
+                    },
+                    Some(c) => buffer.push(c),
+                    None => bail!("unclosed string literal"),
+                }
+            }
+        }
+        Some(c) if c.is_ascii_whitespace() => {
+            let mut buffer = String::new();
+            buffer.push(c);
+            loop {
+                match peek!() {
+                    Some(c) if c.is_ascii_whitespace() => {
+                        buffer.push(*c);
+                        input.next();
+                    }
+                    _ => break,
+                }
+            }
+            Some(Token::Whitespace(buffer))
+        }
+        Some(c) => {
+            let mut buffer = String::new();
+            buffer.push(c);
+            loop {
+                match peek!() {
+                    Some('"') => break,
+                    Some(c) if c.is_ascii_whitespace() => break,
+                    None => break,
+
+                    Some(c) => {
+                        buffer.push(*c);
+                        input.next();
+                    }
+                }
+            }
+            Some(Token::String(buffer))
+        }
+        None => None,
+    };
+    Ok(res)
+}
+
+fn tokenize_patterns(s: &str) -> Result<Vec<Sp<Token>>> {
+    let mut input = s.char_indices().peekable();
+    let mut tokens = vec![];
+    loop {
+        let start = match input.peek() {
+            Some((idx, _)) => *idx,
+            None => break,
+        };
+        let tok = match next_token(&mut input)? {
+            Some(tok) => tok,
+            None => break,
+        };
+        let end = match input.peek() {
+            Some((idx, _)) => *idx,
+            None => s.len(),
+        };
+        tokens.push(Sp::new(tok, start, end));
+    }
+    Ok(tokens)
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum Directive {
+    Check(String),
+    Not(String),
+}
+
+impl Directive {
+    pub fn is_positive(&self) -> bool {
+        match self {
+            Self::Check(_) => true,
+            Self::Not(_) => false,
+        }
+    }
+
+    pub fn is_negative(&self) -> bool {
+        match self {
+            Self::Check(_) => false,
+            Self::Not(_) => true,
+        }
+    }
+
+    pub fn pattern_str(&self) -> &str {
+        match self {
+            Self::Check(s) | Self::Not(s) => &s,
+        }
+    }
+
+    pub fn parse_line(s: &str) -> Result<Vec<Sp<Directive>>> {
+        // TODO: rewrite the way the offset is counted.
+        let mut offset = 0;
+
+        macro_rules! trim {
+            ($s: expr) => {{
+                let s = $s;
+                let mut iter = s.char_indices();
+                loop {
```

Comment:
consider `while let`

Reply:
Tried `while let` & `for`. They didn't work in this case as "break <value>;" is only allowed in a loop. I'd be tempted to keep this code unmodified for now, but will get rid of them altogether in a later cleanup.